lib/github/ldap/member_search/recursive.rb

```diff
@@ -34,73 +34,72 @@ def initialize(ldap, options = {})
+            groups.concat ldap.search(base: dn, scope: Net::LDAP::SearchScope_BaseObject, attributes: attrs, filter: filter)
+            searched << dn
           end
 
-          # descend to `depth` levels, at most
-          depth.times do |n|
-            # find every (new, unique) member entry
-            depth_subentries = entries.each_with_object([]) do |entry, depth_entries|
-              submembers = entry["member"]
+          # track found groups
+          groups.each { |g| found[g.dn] = g }
 
-              # skip any members we've already found
-              submembers.reject! { |dn| found.key?(dn) }
+          # recursively find subgroups
+          unless groups.empty?
+            depth.times do |n|
+              # pull out subgroups' member DNs to search through
+              sub_dns = groups.each_with_object([]) do |subgroup, sub_dns|
+                sub_dns.concat member_dns(subgroup)
+              end
 
-              # find members of subgroup, including subgroups (N queries)
-              subentries = member_entries(entry)
-              next if subentries.empty?
+              # give up if there's nothing else to search for
+              break if sub_dns.empty?
 
-              # track found subentries
-              subentries.each { |entry| found[entry.dn] = entry }
+              # filter out if already searched for
+              sub_dns.reject! { |dn| searched.include?(dn) }
 
-              # collect all entries for this depth
-              depth_entries.concat subentries
-            end
+              # search for subgroups
+              subgroups = sub_dns.each_with_object([]) do |dn, subgroups|
+                subgroups.concat ldap.search(base: dn, scope: Net::LDAP::SearchScope_BaseObject, attributes: attrs, filter: filter)
+                searched << dn
+              end
 
-            # stop if there are no more subgroups to search
-            break if depth_subentries.empty?
+              break if subgroups.empty?
 
-            # go one level deeper
-            entries = depth_subentries
-          end
+              # track found groups
+              subgroups.each { |g| found[g.dn] = g }
 
-          # return all found entries
-          found.values
-        end
+              # descend another level
+              groups = subgroups
+            end
+          end
 
-        # Internal: Fetch member entries, including subgroups, for the given
-        # entry.
-        #
-        # Returns an Array of Net::LDAP::Entry objects.
-        def member_entries(entry)
-          entries = []
-          dns     = member_dns(entry)
-          uids    = member_uids(entry)
+          # pull member DNs, discarding dupes and subgroup DNs
+          member_dns = found.values.each_with_object([]) do |group, member_dns|
+            entries << group
+            member_dns.concat member_dns(group)
+          end.uniq.reject { |dn| found.key?(dn) }
```

Comment:
Could you accomplish this with a `select` rather than an `each_with_object`? Doing the rejecting while iterating would save a pass.